### PR TITLE
add rules

### PR DESCRIPTION
**Pull request by charleenfei**
<This is intended as a guide when describing-reviewing pull requests for go-centrifuge>
Fixs #800.

### CODE REVIEW

File: documents/coredocument.go [hunk 1]
```diff
@@ -6,11 +6,12 @@ import (
 	"fmt"
 	"strings"
 
+	"github.com/centrifuge/go-centrifuge/utils"
+
```
Comment (vedhavyas) [inline documents/coredocument.go]: fix the import spacing

File: documents/coredocument.go [hunk 2]
```diff
@@ -203,32 +206,26 @@ func (cd *CoreDocument) PrepareNewVersion(collaborators []string, initSalts bool
...
-func (cd *CoreDocument) addCollaboratorsToReadSignRules(collaborators []identity.DID) {
+// newRole returns a new role with random role key
+func newRole() *coredocumentpb.Role {
+	return &coredocumentpb.Role{RoleKey: utils.RandomSlice(idSize)}
+}
+
+// addCollaboratorsToNewRole creates a new Role and adds the given collaborators to this Role.
+// The Role is then returned.
+// The operation returns a nil Role if no collaborators are provided.
+func (cd *CoreDocument) addCollaboratorsToNewRole(collaborators []identity.DID) *coredocumentpb.Role {
```
Comment (vedhavyas) [inline documents/coredocument.go]: This is apure function. Lets not make it part of the COreDocument. I would prefer it if you create a new function called `newRoleWithCollaborators` instead

File: documents/write_acls.go [hunk 3]
```diff
@@ -3,6 +3,10 @@ package documents
 import (
 	"bytes"
 
+	"github.com/centrifuge/centrifuge-protobufs/gen/go/coredocument"
+	"github.com/centrifuge/go-centrifuge/identity"
+	"github.com/centrifuge/go-centrifuge/utils"
+
```
Comment (vedhavyas) [inline documents/write_acls.go]: import spacing

File: documents/write_acls.go [hunk 4]
```diff
@@ -84,3 +88,39 @@ func newChangedField(p proofs.Property, leaf *proofs.LeafNode, old bool) changed
...
+}
+
+// addCollaboratorsToTransitionRules adds the given collaborators to a new transition rule which defaults to
+// granting edit capability over all fields of the document.
+func (cd *CoreDocument) addCollaboratorsToTransitionRules(collaborators []identity.DID) {
+	role := cd.addCollaboratorsToNewRole(collaborators)
+	if role == nil {
+		return
+	}
+	cd.addNewTransitionRule(role, coredocumentpb.FieldMatchType_FIELD_MATCH_TYPE_PREFIX, []byte(compactProperties(CDTreePrefix)), coredocumentpb.TransitionAction_TRANSITION_ACTION_EDIT)
```
Comment (vedhavyas) [inline documents/write_acls.go]: you are only passing CDtree alone. Init should take specific document prefix and add a new transition rule for that as well

File: documents/write_acls_test.go [hunk 5]
```diff
@@ -4,11 +4,12 @@ package documents
 
 import (
 	"crypto/sha256"
-	"fmt"
 	"reflect"
 	"testing"
 	"time"
 
+	"github.com/centrifuge/go-centrifuge/identity"
+
```
Comment (vedhavyas) [inline documents/write_acls_test.go]: same as above

File: documents/write_acls.go (hunk 4 above)
Comment (vedhavyas) [reply documents/write_acls.go]: Else, no one will ever make any changes to the document underneath

File: documents/coredocument.go [hunk 6]
```diff
@@ -83,8 +83,9 @@ func NewCoreDocumentFromProtobuf(cd coredocumentpb.CoreDocument) *CoreDocument {
 	return &CoreDocument{Document: cd}
 }
 
-// NewCoreDocumentWithCollaborators generates new core Document, adds collaborators, adds read rules and fills salts
-func NewCoreDocumentWithCollaborators(collaborators []string) (*CoreDocument, error) {
+// NewCoreDocumentWithCollaborators generates new core Document with a document type specified by the prefix: CdTreePrefix, po, or invoice.
```
Comment (vedhavyas) [inline documents/coredocument.go]:
```suggestion
// NewCoreDocumentWithCollaborators generates new core Document with a document type specified by the prefix: Purchase order or invoice.
```

File: documents/coredocument.go [hunk 7]
```diff
@@ -83,8 +83,9 @@ func NewCoreDocumentFromProtobuf(cd coredocumentpb.CoreDocument) *CoreDocument {
...
+// It then adds collaborators, adds read rules and fills salts.
+func NewCoreDocumentWithCollaborators(collaborators []string, compactPrefix []byte) (*CoreDocument, error) {
```
Comment (vedhavyas) [inline documents/coredocument.go]:
```suggestion
func NewCoreDocumentWithCollaborators(collaborators []string, documentPrefix []byte) (*CoreDocument, error) {
```

File: documents/coredocument.go [hunk 8]
```diff
@@ -156,7 +158,7 @@ func (cd *CoreDocument) setSalts() error {
 
 // PrepareNewVersion prepares the next version of the CoreDocument
 // if initSalts is true, salts will be generated for new version.
-func (cd *CoreDocument) PrepareNewVersion(collaborators []string, initSalts bool) (*CoreDocument, error) {
+func (cd *CoreDocument) PrepareNewVersion(collaborators []string, initSalts bool, compactPrefix []byte) (*CoreDocument, error) {
```
Comment (vedhavyas) [inline documents/coredocument.go]:
```suggestion
func (cd *CoreDocument) PrepareNewVersion(collaborators []string, initSalts bool, documentPrefix []byte) (*CoreDocument, error) {
```

File: documents/coredocument.go [hunk 9]
```diff
@@ -203,32 +206,26 @@ func (cd *CoreDocument) PrepareNewVersion(collaborators []string, initSalts bool
...
-// addCollaboratorsToReadSignRules adds the given collaborators to a new read rule with READ_SIGN capability.
-// The operation is no-op if no collaborators is provided.
-// The operation is not idempotent. So calling twice with same accounts will lead to read rules duplication.
-func (cd *CoreDocument) addCollaboratorsToReadSignRules(collaborators []identity.DID) {
+// newRole returns a new role with random role key
+func newRole() *coredocumentpb.Role {
+	return &coredocumentpb.Role{RoleKey: utils.RandomSlice(idSize)}
+}
+
+// addCollaboratorsToNewRole creates a new Role and adds the given collaborators to this Role.
```
Comment (vedhavyas) [inline documents/coredocument.go]:
```suggestion
// newRoleWithCollaborators creates a new Role and adds the given collaborators to this Role.
```

File: documents/write_acls.go [hunk 10]
```diff
@@ -84,3 +87,40 @@ func newChangedField(p proofs.Property, leaf *proofs.LeafNode, old bool) changed
 	cf.new = v
 	return cf
 }
+
+// initTransitionRules initiates the transition rules for a given CoreDocumentModel.
```
Comment (vedhavyas) [inline documents/write_acls.go]:
```suggestion
// initTransitionRules initiates the transition rules for a given Core Document.
```

File: documents/write_acls.go [hunk 11]
```diff
@@ -84,3 +87,40 @@ func newChangedField(p proofs.Property, leaf *proofs.LeafNode, old bool) changed
...
+// Collaborators are given default edit capability over all fields of the CoreDocument.
```
Comment (vedhavyas) [inline documents/write_acls.go]:
```suggestion
// Collaborators are given default edit capability over all fields of the CoreDocument and underlying documents like invoice or purchase order.
```

File: documents/write_acls.go [hunk 12]
```diff
@@ -84,3 +87,40 @@ func newChangedField(p proofs.Property, leaf *proofs.LeafNode, old bool) changed
...
+// if the rules are created already, this is a no-op.
+// if collaborators are empty, it is a no-op
+func (cd *CoreDocument) initTransitionRules(collaborators []identity.DID, compactPrefix []byte) {
```
Comment (vedhavyas) [inline documents/write_acls.go]:
```suggestion
func (cd *CoreDocument) initTransitionRules(collaborators []identity.DID, documentPrefix []byte) {
```

File: documents/write_acls.go [hunk 13]
```diff
@@ -84,3 +87,40 @@ func newChangedField(p proofs.Property, leaf *proofs.LeafNode, old bool) changed
...
+		return
+	}
+	cd.addNewTransitionRule(role, coredocumentpb.FieldMatchType_FIELD_MATCH_TYPE_PREFIX, compactProperties(CDTreePrefix), coredocumentpb.TransitionAction_TRANSITION_ACTION_EDIT)
+	cd.addNewTransitionRule(role, coredocumentpb.FieldMatchType_FIELD_MATCH_TYPE_PREFIX, compactPrefix, coredocumentpb.TransitionAction_TRANSITION_ACTION_EDIT)
+}
+
+// addNewTransitionRule creates a new transition rule with the given parameters.
+func (cd *CoreDocument) addNewTransitionRule(role *coredocumentpb.Role, matchType coredocumentpb.FieldMatchType, field []byte, action coredocumentpb.TransitionAction) {
+	cd.Document.Roles = append(cd.Document.Roles, role)
+	rule := new(coredocumentpb.TransitionRule)
```
Comment (vedhavyas) [inline documents/write_acls.go]: use a struct initialisation here

```
rule := &coredocumentpb.TransitionRule{
	...
}
```

File: documents/write_acls_test.go [hunk 14]
```diff
@@ -66,24 +66,30 @@ func TestWriteACLs_getChangedFields_with_core_document(t *testing.T) {
...
+	// transition_rules.RuleKey
+	// (transition_rules.Roles
+	// transition_rules.MatchType
+	// transition_rules.Action
+	// transition_rules.Field) x 2
 	oldTree := getTree(t, &doc.Document)
 	newTree := getTree(t, &ndoc.Document)
 	cf := getChangedFields(oldTree, newTree, proofs.DefaultSaltsLengthSuffix)
-	assert.Len(t, cf, 9)
+	assert.Len(t, cf, 20)
```
Comment (vedhavyas) [inline documents/write_acls_test.go]: how come we increased the leaves by 11 here?

File: p2p/receiver/handler_test.go [hunk 15]
```diff
@@ -178,7 +178,7 @@ func TestHandler_HandleInterceptor_NilDocument(t *testing.T) {
 
 func TestHandler_HandleInterceptor_getServiceAndModel_fail(t *testing.T) {
 	ctx := testingconfig.CreateAccountContext(t, cfg)
-	cd, err := documents.NewCoreDocumentWithCollaborators(nil)
+	cd, err := documents.NewCoreDocumentWithCollaborators(nil, []byte(documents.CDTreePrefix))
```
Comment (vedhavyas) [inline p2p/receiver/handler_test.go]:
```suggestion
	cd, err := documents.NewCoreDocumentWithCollaborators(nil, nil)
```

Comment (vedhavyas) [reply p2p/receiver/handler_test.go]: by passing the CDPRefix here, you are creating two rules with same prefix

File: documents/write_acls_test.go (hunk 14 above)
Comment (charleenfei) [reply documents/write_acls_test.go]: there are 10 new fields from the two new transition rules, 2 new files from the new read rule, and two new roles. this is in addition to the 6 changed fields for the non-rule/role related changes, adding up to 20.

File: documents/read_acls.go [hunk 16]
```diff
@@ -33,6 +33,27 @@ func (cd *CoreDocument) initReadRules(collaborators []identity.DID) {
...
+	cd.addNewReadRule(role, coredocumentpb.Action_ACTION_READ_SIGN)
+}
+
+// addNewReadRule creates a new read rule as per the role and action.
+func (cd *CoreDocument) addNewReadRule(role *coredocumentpb.Role, action coredocumentpb.Action) {
+	cd.Document.Roles = append(cd.Document.Roles, role)
+	rule := &coredocumentpb.ReadRule{
+		Action: action,
+		Roles:  [][]byte{role.RoleKey},
+	}
```
Comment (vedhavyas) [inline documents/read_acls.go]: perfecto!

**Pull request merged**